src/server/getters/database.ts

```diff
@@ -214,27 +214,55 @@ export async function queryTradingHistoryParams(db: Knex, params: t.TypeOf<typeo
       params.limit,
       params.offset,
       params.ignoreSelfTrades,
-      (err: Error|null, userTradingHistory?: Array<TradingHistoryRow>): void => {
+      (err: Error | null, userTradingHistory?: Array<TradingHistoryRow>): void => {
         if (err) return reject(err);
         resolve(userTradingHistory);
       });
   });
 }
 
+/*
+  * Given a random timestamp will return q query for the apropos block number for the time period in question.
+  * If a zero is given, will return the highest or lowest blocknumber we have in the db.
+  * If the time period is out of range will return a query that will return noting (technically an error).
+  * @param {db} knex handle
+  * @param {timestamp} a timestamp for which you want a block or zero
+  * @param {prefix} 'start' or 'end' for weather you want a startBlock, startTime or endBlock, endTime
+  * @returns {Promise<Knex>} the query
+  */
+export function blockNumberForTimestampQuery(db: Knex, timestamp: number, prefix: string): Knex.QueryBuilder {
+  // This constant is related to the maximum recorded time between blocks so that
+  // our query will pick up a blockNumber for the timestamp in unusual conditions.
+  const MAX_BLOCKTIME_DELTA = 29;
+
+  if (timestamp === 0) {
+    return db.select(`blockNumber as ${prefix}Block`, `timestamp as ${prefix}Time`)
+      .from("blocks")
+      .orderBy("blockNumber", prefix === "start" ? "asc" : "desc")
+      .limit(1);
+  }
+
+  return db.select(`blockNumber as ${prefix}Block`, `timestamp as ${prefix}Time`)
```

Comment:
Grabs the block that encompasses the timestamp whether the block is mined 2 days from now or on time in 14.5 seconds (or less).